### ISSUE TITLE
Guidance for lossy 16bit compression

### ISSUE THREAD

**Issue by tbonfort** [libjxl/libjxl#1763]
I'm implementing support for lossy 16bit RGB compression with JXL and am encountering an issue. However I can reproduce my issue with cjxl.

Test image (16bit png, so doesn't render in gihub):
![extract](https://user-images.githubusercontent.com/358469/191227466-2cde4a16-9945-4464-b5c0-c2a622b0f1ec.png)

Test setup: `cjxl extract.png extract.jxl` , i.e. expecting compression with distance=1, visually lossless

Expected result: visually lossless jxl
Observed result: 1.5kb jxl image, severlely lossy and barely recognisable

Attached screenshots reconverted to 8bits for visualization:
Original:
![Screenshot from 2022-09-20 11-57-12](https://user-images.githubusercontent.com/358469/191228382-625919f7-b0e1-4fc4-ab94-d05139ce4dcf.png)
JXL:
![Screenshot from 2022-09-20 11-57-17](https://user-images.githubusercontent.com/358469/191228499-b5ce2bdc-6a81-46a7-b18f-c46e7d5f3e1c.png)

Is this configuration expected to be supported? If so how?

**Comment by jyrkialakuijala** [comment]
Lots of guesswork in my response, so feel free to ignore. Consider it as an early uneducated response:

Usually 8 bit to 16 bit conversion is done by multiplying the 8 bit values by 257 -- not just writing zeros into the higher bits.

This is bit of a stretch but would it work if you add --intensity_target=15893440 to cjxl (perhaps there is a max that limits this)

(15893440 would be from 80 for srgb intensity, then multiplied with 256**2.2 more for the information being in the lower 8 bits only)

**Comment by jonsneyers** [comment]
Since the png looks black, I assume you only use the low part of the range — e.g. the actual data is 10-bit, and you encode white as 1023 instead of 65535 when representing it in 16 bits. This is not how you're supposed to represent 10-bit data in PNG, so both cjxl and any png viewer will get it wrong.
Two options: 1) use 10-bit ppm as input instead, 2) make a 16-bit png with padding in the low bits instead of padding in the high bits.

Basically the poor visual result is caused by making libjxl think that the image looks exactly like how your browser will render that png: extremely dark, e.g. if it's 10-bit it's as if all colors are between 8-bit rgb #000000 and #030303. This is obviously not how you intend to view the image.

**Comment by tbonfort** [comment]
Thanks Jon for the insights.
> Since the png looks black, I assume you only use the low part of the range — e.g. the actual data is 10-bit, and you encode white as 1023 instead of 65535 when representing it in 16 bits

Yes that is correct

The png example with cjxl was just an example, my end use-case is to use the libjxl api to encode 16bit data but where the actual range of the data is less than 16 bits.
I'm currently using the encoder with `JxlPixelFormat.data_type=JXL_TYPE_UINT16` and `JxlBasicInfo.bits_per_sample=11` but that is not giving "correct" results. Could you please advise as to how I should be handling this use case?

**Comment by jonsneyers** [comment]
At the moment I don't think we have an API that can take 'uint11' input or produce such output. Maybe we should change that, because it would indeed be more convenient for your use case.

For now, you'll have to manually adjust your buffer to be using the full 16-bit range (i.e. multiply every value by 65535/2047, or do something like `x = (x << 5) | (x >> 6)`. It will still encode it as an 11-bit image if you set the BasicInfo that way, i.e. in lossless mode the 5 lsb will be ignored, and in lossy mode the internal representation is not integer-based anyway so the bit depth is just a header field.

@mo271 This is something we should probably make more convenient in the API, i.e. have a way to use UINT16 buffers where the range is not 0..65535 but something else. This is also needed to make cjxl/djxl handle PPM correctly, which supports arbitrary bitdepths with padding in the high bits and not in the low bits like in PNG. I think we currently misinterpret e.g. 10-bit PPMs as input to cjxl for the same reason as what @tbonfort is experiencing here. Since we're converting to/from float internally anyway, this is just a matter of using a different multiplier for us, while for API users it means they have to convert their buffers from msb-padded to lsb-padded.

**Comment by mo271** [comment]
Adding @szabadka to the conversation, having added https://github.com/libjxl/libjxl/commit/da5d05e35a1d9f9564cb3af0d8389e7b9c1eb564 recently.

**Comment by szabadka** [comment]
I am currently working on this.

As a first step, I will add a testing framework that runs the PackedPixelFile -> Encoder API -> JXL bitstream -> Decoder API -> PackedPixelFile roundtrip. This way we can test end-to-end what happens with different bits-depth/data-type combinations. 
[Currently we only test the CodecInOut -> JXL bitstream -> Decoder API -> PackedPixelFile -> CodecInOut roundtrip].

Then I am planning to add the not-full-bitdepth uint16 input/output feature in some way to the encoder and decoder API.

**Comment by tbonfort** [comment]
From my perspective, where I'm integrating jxl as a codec inside libtiff, what would simplify my life more than high vs low bit padding, would be to support a buffer with no padding at all, as that is what is produced by libtiff internals.
e.g. for a buffer of 64 uint16 pixels with only 12 significant bits, the buffer size is 64*12/8 = 96 bytes long, not 128. (This does of course imply that the length of the provided buffer is a multiple of the number of significant bits and a multiple of 8)

**Comment by mo271** [comment]
Just out of interest, how do you "reconverted to 8bits for visualization" in the original post? If I just do linear scaling assuming this is 11bit data it does not look the same. Is some color profile applied here?

**Comment by jonsneyers** [comment]
> From my perspective, where I'm integrating jxl as a codec inside libtiff, what would simplify my life more than high vs low bit padding, would be to support a buffer with no padding at all, as that is what is produced by libtiff internals. e.g. for a buffer of 64 uint16 pixels with only 12 significant bits, the buffer size is 64*12/8 = 96 bytes long, not 128. (This does of course imply that the length of the provided buffer is a multiple of the number of significant bits and a multiple of 8)

Ah, that's also interesting. In particular the special cases for 1,2,4 bit buffers would also be convenient and it would be a useful reduction in buffer sizes compared to using uint8.

So basically the feature requests here (and in some other open issues) are to add ways to get/set buffers using:

- bit-packing (possibly this can be done with different endiannesses too) — convenient for the libtiff integration
- padding in the high bits instead of in the low bits — convenient for reading/writing ppm that is not 8 or 16-bit
- planar R,G,B instead of interleaved
- other interleaved orderings, e.g. BGR — convenient for opencv, for example
- interleaved CMYK — convenient for anything that deals with CMYK
- possibly interleaved >4 channel, e.g. CMYKA — might be convenient for Krita and ImageMagick?
- interleaved ignored channels, e.g. RGBX can be useful (where X is just ignored)

**Comment by tbonfort** [comment]
> Just out of interest, how do you "reconverted to 8bits for visualization" in the original post? If I just do linear scaling assuming this is 11bit data it does not look the same. Is some color profile applied here?

Sorry I don't remember the exact parameters I used for those screengrabs, that where done in a ui (namely qgis) where I can manually set the scaling for each channel (e.g scale [200-1200] to [0-255]).

**Comment by mo271** [comment]
Is there still something to do here?
See also https://github.com/libjxl/libjxl/issues/3787

**Comment by jonsneyers** [comment]
Regarding the original question: this is resolved now, applications can use `JXL_BIT_DEPTH_FROM_CODESTREAM` if they want to use an intermediate bit depth (one that is not a multiple of 8) with padding in the msb instead of in the lsb.

What is not addressed yet is support for bit-packed formats, support for planar buffers (separate R, G, B planes instead of interleaving them), and different interleaving orders (e.g. BGR instead of RGB). This requires generalizing `JxlPixelFormat` and perhaps adding functions like `JxlDecoderSetExtraChannelBuffer` that allow you to set buffers for single channels of the main color image.

**Comment by mo271** [comment]
ok, so what remains seems to be https://github.com/libjxl/libjxl/issues/1542

**Comment by jonnyawsom3** [comment]
Going to be marking future 16-bit padding issues as duplicates of this one (Technically #314 was the original, but this has more info). Should it be re-opened? There seems to be a non-insignificant amount of cases where pre-processing isn't viable.
